### PR TITLE
improved mixture

### PR DESCRIPTION
**Pull request by matbesancon**
The performance of mixture models is terrible at the moment, as proved in the added `perf/mixtures.jl` file.

Here are the result:
```julia
julia> include("mixtures.jl")
[ Info: sampling 0.5331830160438613
[ Info: evaluate_manual_pdf
  75.281 ns (3 allocations: 80 bytes)
[ Info: current_master
  470.505 ns (15 allocations: 272 bytes)
[ Info: improved_version
  553.541 ns (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  584.072 ns (9 allocations: 192 bytes)
[ Info: improved_zipped
  377.395 ns (5 allocations: 112 bytes)
[ Info: improved_indexed
  263.629 ns (5 allocations: 112 bytes)
[ Info: sampling 0.4540291355871424
[ Info: evaluate_manual_pdf
  75.168 ns (3 allocations: 80 bytes)
[ Info: current_master
  470.010 ns (15 allocations: 272 bytes)
[ Info: improved_version
  569.508 ns (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  599.706 ns (9 allocations: 192 bytes)
[ Info: improved_zipped
  371.519 ns (5 allocations: 112 bytes)
[ Info: improved_indexed
  267.948 ns (5 allocations: 112 bytes)
[ Info: sampling 0.017686826714964354
[ Info: evaluate_manual_pdf
  69.143 ns (3 allocations: 80 bytes)
[ Info: current_master
  464.056 ns (15 allocations: 272 bytes)
[ Info: improved_version
  560.951 ns (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  577.039 ns (9 allocations: 192 bytes)
[ Info: improved_zipped
  358.352 ns (5 allocations: 112 bytes)
[ Info: improved_indexed
  264.006 ns (5 allocations: 112 bytes)
[ Info: sampling 0.17293302893695128
[ Info: evaluate_manual_pdf
  69.242 ns (3 allocations: 80 bytes)
[ Info: current_master
  464.408 ns (15 allocations: 272 bytes)
[ Info: improved_version
  564.859 ns (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  584.464 ns (9 allocations: 192 bytes)
[ Info: improved_zipped
  365.293 ns (5 allocations: 112 bytes)
[ Info: improved_indexed
  263.831 ns (5 allocations: 112 bytes)
[ Info: sampling 0.9589258763297348
[ Info: evaluate_manual_pdf
  76.720 ns (3 allocations: 80 bytes)
[ Info: current_master
  455.234 ns (15 allocations: 272 bytes)
[ Info: improved_version
  583.144 ns (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  586.072 ns (9 allocations: 192 bytes)
[ Info: improved_zipped
  371.335 ns (5 allocations: 112 bytes)
[ Info: improved_indexed
  268.012 ns (5 allocations: 112 bytes)
[ Info: sampling 0.5331830160438613
[ Info: evaluate_manual_pdf
  15.936 μs (3 allocations: 80 bytes)
[ Info: current_master
  129.241 μs (5980 allocations: 109.05 KiB)
[ Info: improved_version
  15.914 μs (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  15.928 μs (9 allocations: 192 bytes)
[ Info: improved_zipped
  16.240 μs (5 allocations: 112 bytes)
[ Info: improved_indexed
  15.569 μs (5 allocations: 112 bytes)
[ Info: sampling 0.4540291355871424
[ Info: evaluate_manual_pdf
  15.991 μs (3 allocations: 80 bytes)
[ Info: current_master
  129.735 μs (5980 allocations: 109.05 KiB)
[ Info: improved_version
  15.903 μs (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  15.809 μs (9 allocations: 192 bytes)
[ Info: improved_zipped
  16.216 μs (5 allocations: 112 bytes)
[ Info: improved_indexed
  15.667 μs (5 allocations: 112 bytes)
[ Info: sampling 0.017686826714964354
[ Info: evaluate_manual_pdf
  18.499 μs (3 allocations: 80 bytes)
[ Info: current_master
  133.670 μs (5980 allocations: 109.05 KiB)
[ Info: improved_version
  18.441 μs (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  18.231 μs (9 allocations: 192 bytes)
[ Info: improved_zipped
  19.053 μs (5 allocations: 112 bytes)
[ Info: improved_indexed
  18.187 μs (5 allocations: 112 bytes)
[ Info: sampling 0.17293302893695128
[ Info: evaluate_manual_pdf
  16.944 μs (3 allocations: 80 bytes)
[ Info: current_master
  132.169 μs (5980 allocations: 109.05 KiB)
[ Info: improved_version
  16.825 μs (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  16.680 μs (9 allocations: 192 bytes)
[ Info: improved_zipped
  17.417 μs (5 allocations: 112 bytes)
[ Info: improved_indexed
  16.610 μs (5 allocations: 112 bytes)
[ Info: sampling 0.9589258763297348
[ Info: evaluate_manual_pdf
  18.232 μs (3 allocations: 80 bytes)
[ Info: current_master
  133.976 μs (5980 allocations: 109.05 KiB)
[ Info: improved_version
  18.175 μs (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  18.163 μs (9 allocations: 192 bytes)
[ Info: improved_zipped
  18.823 μs (5 allocations: 112 bytes)
[ Info: improved_indexed
  17.981 μs (5 allocations: 112 bytes)
```

### CODE REVIEW

**Comment by matbesancon** [comment]
The implementation chosen is `improved_indexed`.

File: perf/mixtures.jl [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    Normal(-1.0, 0.3),
+    Normal(0.0, 0.5),
+    Normal(3.0, 1.0),
+]
+
+priors = [0.25, 0.25, 0.5]
+
+gmm_normal = MixtureModel(distributions, priors)
+
+@info "Small Gausian mixture"
```
Comment (nalimilan) [inline perf/mixtures.jl]:
```suggestion
@info "Small Gaussian mixture"
```

File: src/mixtures/mixturemodel.jl [hunk 2]
```diff
@@ -48,7 +48,7 @@ component_type(d::AbstractMixtureModel{VF,VS,C}) where {VF,VS,C} = C
 
 Get a list of components of the mixture model `d`.
 """
-components(d::AbstractMixtureModel)
+components(d::AbstractMixtureModel) = [component(d, k) for k in Base.OneTo(ncomponents(d))]
```
Comment (nalimilan) [inline src/mixtures/mixturemodel.jl]: I don't think `Base.OneTo` is really useful here, is it?
```suggestion
components(d::AbstractMixtureModel) = [component(d, k) for k in 1:ncomponents(d)]
```

**Comment by mschauer** [comment]
The current implementation has the advantage that NaN values with zero probabilities do not pollute the result. That would correspond to something like 
```
sum(p > 0 ? p*pdf(c, x) : zero(p) for (p, c) in zip(ps, cs))
```

**Comment by matbesancon** [comment]
> The current implementation has the advantage that NaN values with zero probabilities do not pollute the result

So we would prefer replacing NaN with zero instead of propagating them?

**Comment by matbesancon** [comment]
> Note that the components fallback will return a vector which can be modified by the caller, but AFAICT that's not the case of the MixtureModel specialization, which returns an internal field without copying.

Yes I think we should default to not copying when possible, the fallback is inefficient and should be overridden by subtypes of AbstractMixture, even though it's not a drama

**Comment by mschauer** [comment]
> So we would prefer replacing NaN with zero instead of propagating them?

These expressions are principle integrals, where we want to set 0⋅∞ = 0, similarly as it is the convention in measure theory https://en.wikipedia.org/wiki/Lebesgue_integration#Simple_functions

**Comment by matbesancon** [comment]
With the additional check, it gets harder to get back to the same perf levels for small arrays

**Comment by matbesancon** [comment]
The best result I get is with a boolean product, which avoids branching:
```julia
function indexed_boolprod(d, x)
    ps = probs(d)
    cs = components(d)
    @inbounds sum((ps[i] > 0) * ps[i] * pdf(cs[i], x) for i in eachindex(ps))
end
```

**Comment by matbesancon** [comment]
An example from each now:
```
[ Info: Small Gausian mixture
[ Info: sampling 0.5331830160438613
[ Info: evaluate_manual_pdf
  77.130 ns (3 allocations: 80 bytes)
[ Info: current_master
  471.526 ns (15 allocations: 272 bytes)
[ Info: improved_version
  791.863 ns (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  685.993 ns (9 allocations: 192 bytes)
[ Info: ifelse_version
  869.947 ns (13 allocations: 256 bytes)
[ Info: forloop
  605.726 ns (24 allocations: 464 bytes)
[ Info: indexed_sum_comp
  826.403 ns (12 allocations: 240 bytes)
[ Info: indexed_boolprod
  267.362 ns (5 allocations: 112 bytes)
[ Info: sumcomp_cond
  883.531 ns (12 allocations: 240 bytes)

[ Info: Large Gausian mixture
[ Info: sampling 0.5331830160438613
[ Info: evaluate_manual_pdf
  16.026 μs (3 allocations: 80 bytes)
[ Info: current_master
  125.115 μs (5980 allocations: 109.05 KiB)
[ Info: improved_version
  15.850 μs (9 allocations: 192 bytes)
[ Info: improved_no_inbound
  19.131 μs (9 allocations: 192 bytes)
[ Info: ifelse_version
  85.996 μs (2007 allocations: 31.41 KiB)
[ Info: forloop
  163.018 μs (8978 allocations: 171.52 KiB)
[ Info: indexed_sum_comp
  16.894 μs (12 allocations: 240 bytes)
[ Info: indexed_boolprod
  15.746 μs (5 allocations: 112 bytes)
[ Info: sumcomp_cond
  16.925 μs (12 allocations: 240 bytes)

[ Info: Heterogeneous distributions
[ Info: sampling 0.5331830160438613
[ Info: evaluate_manual_pdf
  416.438 μs (10007 allocations: 187.66 KiB)
[ Info: current_master
  298.842 μs (14980 allocations: 265.30 KiB)
[ Info: improved_version
  174.724 μs (10008 allocations: 156.42 KiB)
[ Info: improved_no_inbound
  178.634 μs (10007 allocations: 156.41 KiB)
[ Info: ifelse_version
  256.861 μs (10007 allocations: 156.41 KiB)
[ Info: forloop
  355.814 μs (16978 allocations: 296.52 KiB)
[ Info: indexed_sum_comp
  171.247 μs (10009 allocations: 156.44 KiB)
[ Info: indexed_boolprod
  169.861 μs (10007 allocations: 156.39 KiB)
[ Info: sumcomp_cond
  171.635 μs (10009 allocations: 156.44 KiB)
```

**Comment by nalimilan** [comment]
Are you sure `@inbounds` has an effect when outside of the generator?

**Comment by matbesancon** [comment]
It's supposed to propagate in the whole expression, I'll run the benchmark on the same function without it

**Comment by matbesancon** [comment]
Last benchmarks with the comparisons:
```
[ Info: indexed_boolprod
  266.676 ns (5 allocations: 112 bytes)
[ Info: indexed_boolprod_noinbound
  275.064 ns (5 allocations: 112 bytes)
[ Info: indexed_boolprod
  266.450 ns (5 allocations: 112 bytes)
[ Info: indexed_boolprod_noinbound
  268.755 ns (5 allocations: 112 bytes)
[ Info: indexed_boolprod
  259.679 ns (5 allocations: 112 bytes)
[ Info: indexed_boolprod_noinbound
  268.375 ns (5 allocations: 112 bytes)

[ Info: indexed_boolprod
  267.171 ns (5 allocations: 112 bytes)
[ Info: indexed_boolprod_noinbound
  273.960 ns (5 allocations: 112 bytes)

# bigger
[ Info: indexed_boolprod
  16.198 μs (5 allocations: 112 bytes)
[ Info: indexed_boolprod_noinbound
  15.781 μs (5 allocations: 112 bytes)

[ Info: indexed_boolprod
  16.346 μs (5 allocations: 112 bytes)
[ Info: indexed_boolprod_noinbound
  15.910 μs (5 allocations: 112 bytes)

[ Info: indexed_boolprod
  18.970 μs (5 allocations: 112 bytes)
[ Info: indexed_boolprod_noinbound
  18.116 μs (5 allocations: 112 bytes)

[ Info: indexed_boolprod
  17.510 μs (5 allocations: 112 bytes)
[ Info: indexed_boolprod_noinbound
  16.866 μs (5 allocations: 112 bytes)
[ Info: indexed_boolprod
  18.813 μs (5 allocations: 112 bytes)
[ Info: indexed_boolprod_noinbound
  17.985 μs (5 allocations: 112 bytes)

# heterogeneous distributions

[ Info: indexed_boolprod
  202.612 μs (12007 allocations: 187.64 KiB)
[ Info: indexed_boolprod_noinbound
  203.257 μs (12007 allocations: 187.64 KiB)
[ Info: indexed_boolprod
  204.411 μs (12007 allocations: 187.64 KiB)
[ Info: indexed_boolprod_noinbound
  201.409 μs (12007 allocations: 187.64 KiB)
[ Info: indexed_boolprod
  201.608 μs (12007 allocations: 187.64 KiB)
[ Info: indexed_boolprod_noinbound
  197.743 μs (12007 allocations: 187.64 KiB)
```

The evidence is not striking except for small indices, but we can remove the inbounds

File: src/mixtures/mixturemodel.jl [hunk 3]
```diff
@@ -48,7 +48,7 @@ component_type(d::AbstractMixtureModel{VF,VS,C}) where {VF,VS,C} = C
 
 Get a list of components of the mixture model `d`.
 """
-components(d::AbstractMixtureModel)
+components(d::AbstractMixtureModel) = [component(d, k) for k in 1:ncomponents(d)]
```
Comment (mschauer) [inline src/mixtures/mixturemodel.jl]: What is this actually for? `component` is - I think - not defined for the `AbstractMixtureModel` and `MixtureModel` already has a method `components`

Comment (matbesancon) [reply src/mixtures/mixturemodel.jl]: From the documentation of `AbstractMixture`:
```julia
  All subtypes of `AbstractMixtureModel` should implement the following methods:
  - ncomponents(d): the number of components
  - component(d, k):  return the k-th component
  - probs(d):       return a vector of prior probabilities over components.
```

Given these basic functions, a fallback for `components` can be defined at the level of AbstractMixture

**Comment by matbesancon** [comment]
I'm adding a test validating the fact that we respect the 0 * NaN = 0 behavior

**Pull request merged**